netsim/augment/links.py

```diff
@@ -288,155 +214,290 @@ def augment_link_prefix(link: Box,pools: typing.List[str],addr_pools: Box,link_p
 
-  for node_if in interfaces:
-    node_if['data'].neighbors = []
-    for remote_if in interfaces:
-      if remote_if['node'] != node_if['node'] or remote_if['data'].ifindex != node_if['data'].ifindex:
-        ngh_data = Box({ 'ifname': remote_if['data'].ifname, 'node': remote_if['node'] })
-        for af in ('ipv4','ipv6'):
-          if af in remote_if['data']:
-            ngh_data[af] = remote_if['data'][af]
-        
-        # List enabled modules that have interface level attributes; copy those attributes too
-        mods_with_ifattr = Box({ m : True for m in ndict[remote_if['node']].get('module',[]) if defaults[m].attributes.get('interface',None) })
-        ifaddr_add_module(ngh_data,remote_if['data'],mods_with_ifattr)
-
-        node_if['data'].neighbors.append(ngh_data)
+IPAM_dispatch: typing.Final[dict] = { 
+    'unnumbered': IPAM_unnumbered,
+    'p2p': IPAM_p2p,
+    'sequential': IPAM_sequential,
+    'id_based': IPAM_id_based
+  }
 
-  if common.debug_active('links'):     # pragma: no cover (debugging)
-    print(f'Final LAN link data: {link}\n')
+"""
+Assign addresses to all interfaces on a link
+
+* Skip if the link has no usable prefix (l2only link)
+* Use IPAM_unnumbered on old-style unnumbered links
+* Figure out allocation policies (based on link type, prefix size, number of interfaces)
+* Execute selected IPAM routing
+"""
+def assign_interface_addresses(link: Box, addr_pools: Box, ndict: Box, defaults: Box) -> None:
+  global IPAM_dispatch
+  pfx_list = link.get('prefix',None)
+  if not pfx_list:
+    return
+
+  if 'unnumbered' in pfx_list:                        # Deal with unnumbered links first
+    for af in ('ipv4','ipv6'):
+      pfx_list.pop(af)                                # Remove AF prefix from unnumbered link
+      IPAM_unnumbered(link,af,None,ndict)
+
+    link.pop('prefix')
+    return
+
+  for af in ('ipv4','ipv6'):
+    if not af in pfx_list:                            # Skip address families not used on the link
+      continue
+
+    if isinstance(pfx_list[af],bool):                 # Unnumbered AF
+      allocation_policy = 'unnumbered'
+      pfx_net = pfx_list[af]
+    else:
+      try:                                            # Parse the AF prefix
+        pfx_net = netaddr.IPNetwork(pfx_list[af])
+      except Exception as ex:                         # Report an error and move on if it cannot be parsed
+        common.error(
+          f'Cannot parse {af} prefix {pfx_list[af]} on link#{link.linkindex}\n' + \
+            common.extra_data_printout(f'{ex}') + '\n' + \
+            common.extra_data_printout(f'{link}'),
+          common.IncorrectValue,
+          'links')
+        continue
+
+      allocation_policy = get_prefix_IPAM_policy(link,pfx_net,ndict)      # get IPAM policy based on prefix and link size
+
+    if allocation_policy == 'error':                                      # Something went wrong, cannot assing IP addresses
```

Comment:
would suggest to include an 'error' policy instead of handling separately here

Reply:
I had that same idea initially, but then it seemed cleaner to handle the error case separately instead of hiding the error message in a routine called through a lookup table.